package dmGameSystemDDF;

import "ddf/ddf_extensions.proto";
import "ddf/ddf_math.proto";

option java_package = "com.dynamo.sprite.proto";
option java_outer_classname = "Sprite";

message SpriteDesc
{
    enum BlendMode
    {
        BLEND_MODE_ALPHA     = 0 [(displayName) = "Alpha"];
        BLEND_MODE_ADD       = 1 [(displayName) = "Add"];
        BLEND_MODE_ADD_ALPHA = 2 [(displayName) = "Add Alpha (Deprecated)"];
        BLEND_MODE_MULT      = 3 [(displayName) = "Multiply"];
    }

    required string tile_set            = 1 [(resource)=true];
    required string default_animation   = 2;
    optional string material            = 3 [(resource)=true, default="/builtins/materials/sprite.material"];
    optional BlendMode blend_mode       = 4 [default = BLEND_MODE_ALPHA];
}

/*# plays a sprite animation
 * <p>
 * Post this message to a sprite-component to make it play an animation from its tile set.
 * </p>
 *
 * @message
 * @name play_animation
 * @param id the id of the animation to play
 * @examples
 * <p>
 * In the examples, it is assumed that the instance of the script has a sprite-component with id "sprite". The sprite itself is assumed to be bound to a tile set with animations "walk" and "jump".
 * <pre>
 * msg.post("#sprite", "play_animation", {id = hash("jump")})
 * </pre>
 * <p>
 * How to sequence two animations together.
 * </p>
 * <pre>
 * function init(self)
 *     -- play jump animation at init
 *     msg.post("#sprite", "play_animation", {id = hash("jump")})
 * end
 *
 * function on_message(self, message_id, message, sender)
 *     -- check for animation done response
 *     if message_id == hash("animation_done") then
 *         -- start the walk animation
 *         msg.post("#sprite", "play_animation", {id = hash("walk")})
 *     end
 * end
 * </pre>
 */
message PlayAnimation
{
    required uint64 id = 1;
}

/*# reports that an animation has completed
 * This message is sent to the sender of a <code>play_animation</code> message when the animation has completed.
 * Note that only animations played either forward or backward once ever completes. See <code>play_animation</code>
 * for more information and examples of how to use this message.
 *
 * @message
 * @name animation_done
 * @param current_tile the current tile of the sprite (number)
 * @param id id of the animation that was completed (hash)
 */
message AnimationDone
{
    required uint32 current_tile = 1;
    required uint64 id = 2;
}

/* Function wrapper documented in gamesys_script.cpp */
message SetFlipHorizontal
{
    required uint32 flip = 1;
}

/* Function wrapper documented in gamesys_script.cpp */
message SetFlipVertical
{
    required uint32 flip = 1;
}

/* Function wrapper documented in gamesys_script.cpp */
message ResetConstant
{
    required uint64         name_hash   = 1;
}

/* Function wrapper documented in gamesys_script.cpp */
message SetScale
{
    required dmMath.Vector3 scale  = 1;
}
